### PR TITLE
Document ad integrations and bring implementation in-line with documentation

### PR DESCRIPTION
**Pull request by cramforce**

### CODE REVIEW

File: 3p/README.md [hunk 1]
```diff
@@ -50,6 +50,8 @@ The better an ad network does on the above requirements, the earlier ads from it
 
 We are also excited to start conversations how modern web tech could improve overall ads latency, memory usage and framerate impact and how AMP's open source model could be used to inject trust through an auditable common code base into the ads ecosystem which would reduce the necessity of redundant metrics collection.
 
+(Further details for ad integrations.)[../ads/README.md]
```
Comment (jmadler) [inline 3p/README.md]: Need to correct the link formatting - I think you've reversed it :)

Comment (jmadler) [inline 3p/README.md]: I'd suggest saying something like "Review the [ads/README] for further details  on ad integration." for better flow

File: 3p/integration.js [hunk 2]
```diff
@@ -88,7 +88,12 @@ window.draw3p = function() {
   window.context.isMaster = window.context.master == window;
   window.context.data = data;
   window.context.noContentAvailable = triggerNoContentAvailable;
-  window.context.updateDimensions = triggerDimensions;
```
Comment (jmadler) [inline 3p/integration.js]: This seems unrelated

File: ads/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+See also our [ad integration guidelines](../3p/README.md#ads).
+
+## Overview
+Ads are just another external resource and must play within the same constraints placed on all resources in AMP. We aim to support a large subset of existing ads with little or no changes to how the integrations work. Our long term goal is to further improve the impact of ads on the user experience through changes across the entire vertical client side stack.
+
+
+## Constraints
+A summary of constraints placed on external resources such as ads in AMP HTML:
+
+- Because AMPs are served on HTTPS and ads cannot be proxied ads must be served on HTTPS.
```
Comment (jmadler) [inline ads/README.md]: comma between "proxied" and "ads"

Comment (jmadler) [inline ads/README.md]: s/on HTTPS/over HTTPS/

File: ads/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+- The size of an ad unit must be static. It must be knowable without fetching the ad and it cannot change at runtime except through https://github.com/ampproject/amphtml/issues/728.
```
Comment (jmadler) [inline ads/README.md]: Suggest changing text to: "except through iframe resizing" (with link to issue), and add a note in that issue to update this doc

File: ads/README.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+- If placing the ad requires running JavaScript (assumed to be true for 100% of ads served through networks), the ad must be placed on an origin different from the AMP document itself.
+Reasons include:
+  - Improved security.
+  - Takes synchronous HTTP requests made by the ad out of the critical path of the primary page.
```
Comment (jmadler) [inline ads/README.md]: critical rendering path

File: ads/README.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+The ad itself is hosted within a document that has an origin different from the primary page.
+
+### Information available to the ad
+We will provide the following information to the ad:
+
+- window.context.referrer contains the origin of the referrer value of the primary document if available.
+- document.referrer will typically contain the URL of the primary document. This may change in the future (See next value for a more reliable method).
+- window.context.location.href contains the sanitized location.href value of the primary document.
+- window.context.canonicalUrl contains the canonical URL of the primary document as defined by its link=canonical tag.
+- ad viewability: Tracked in https://github.com/ampproject/amphtml/issues/720
```
Comment (jmadler) [inline ads/README.md]: note this doc in that issue please

File: ads/README.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+- window.context.noContentAvailable is a function that the ad system can call if the ad slot was not filled. The container page will then react by showing placeholder content (not by collapsing the ad slot).
+
+More information can be provided in a similar fashion if needed.
```
Comment (jmadler) [inline ads/README.md]: Suggest specifying how to request the feature

File: ads/README.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+      type="adtech"
+      src="https://adserver.adtechus.com/addyn/3.0/5280.1/2274008/0/-1/ADTECH;size=300x250;key=plumber;alias=careerbear-ros-middle1;loc=300;;target=_blank;grp=27980912;misc=3767074">
+  </amp-ad>
+```
+
+Note, that the network still needs to be whitelisted and provide a prefix to valid URLs. We may add similar support for ad networks that support loading via an iframe tag.
+
+Technically the `<amp-ad>` tag loads an iframe to a generic bootstrap URL that knows how to render the ad given the parameters to the tag.
+
+### 1st party cookies (not yet implemented)
```
Comment (jmadler) [inline ads/README.md]: Suggest noting this doc in that ticket (and vice versa)

File: 3p/README.md (hunk 1 above)
Comment (cramforce) [inline 3p/README.md]: Done

File: ads/README.md (hunk 3 above)
Comment (cramforce) [inline ads/README.md]: Done

File: ads/README.md (hunk 5 above)
Comment (cramforce) [inline ads/README.md]: Done

File: ads/README.md (hunk 6 above)
Comment (cramforce) [inline ads/README.md]: Done

File: ads/README.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+### Information available to the ad
+We will provide the following information to the ad:
+
+- window.context.referrer contains the origin of the referrer value of the primary document if available.
+- document.referrer will typically contain the URL of the primary document. This may change in the future (See next value for a more reliable method).
+- window.context.location.href contains the sanitized location.href value of the primary document.
+- window.context.canonicalUrl contains the canonical URL of the primary document as defined by its link=canonical tag.
+- ad viewability: Tracked in https://github.com/ampproject/amphtml/issues/720
+- window.context.noContentAvailable is a function that the ad system can call if the ad slot was not filled. The container page will then react by showing placeholder content (not by collapsing the ad slot).
```
Comment (dvoytenko) [inline ads/README.md]: Maybe mention that "iframe resizing" rules will apply or "placeholder" content will be shown. E.g. we can collapse space if outside of the viewport.

File: ads/README.md (hunk 8 above)
Comment (cramforce) [inline ads/README.md]: I'm not yet ready to file it :)

**Comment by dvoytenko** [comment]
LGTM

File: ads/README.md (hunk 4 above)
Comment (cramforce) [inline ads/README.md]: Done

File: ads/README.md (hunk 8 above)
Comment (jmadler) [inline ads/README.md]: Don't forget when you do, then :)

**Comment by jmadler** [comment]
LGTM
